src/Tracking.php

```diff
@@ -84,7 +84,7 @@ public static function maybe_init() {
 		if ( function_exists( 'WC' ) ) {
 
 			if ( ! wp_doing_ajax() ) {
-				add_action( 'wp', array( __CLASS__, 'late_events_handling' ) );
+				add_action( 'wp_head', array( __CLASS__, 'late_events_handling' ), 20 );
```

Comment:
> Since we change the hook did_action( 'wp_head' ) is always going to be truthy

Correct.

> Which means we are always going to delay the tracking of the page visit till the next request from the same user
Not true. I will try to shed some light here:
```
$action = did_action( 'wp_head' ) ? 'print_event' : 'enqueue_event';
```
In `add_event()` the action would be `print_event` which will result in those events being printed immediately. Since we (now) are already in `wp_head` for these events, they will be printed in the `<head>`. 

As implemented, only the events that need to be printed on the next request are stored to the transient. 

The `add_event()` method will either have a `print_event` action which will print the event code on the spot, or have a `enqueue_event` action which will put it in a runtime variable. Events will get stored in that transient if they haven't been printed (so that runtime variable isn't flushed) when reaching `shutdown`. 
So a call to `add_event()` that runs before `wp_head` will enqueue an event, but when `wp_head` does run, it should print the event and clear the queue. 
The only case where an event should end up in the transient is when `add_event()` is called before `wp_head` and `wp_head` is never fired, as in the "Add2Cart with redirect to cart enabled" flow.

Reply:
Thanks for the clarification, I misinterpreted that logic. Because that means with it previously hooking on `wp` it should have always been saving the transient (as long as there was a session). I wasn't seeing that behaviour in my tests, but I suppose that was because `$deferred_conversion_events_transient_key` was still null.

It still leaves me with some questions though.

For an add to cart event we hook into `woocommerce_add_to_cart`, which will always be called before `wp_head`. How are we recording that if we don't have redirect to cart enabled? Does that mean we are still scheduling it for the next request? For example on a single product page when redirect is not enabled (or on a shop page if ajax and redirect are disabled). From my testing it is still being scheduled which makes it seem like the `did_action( 'wp_head' )` is not the right way to detect if we should enqueue or not.

Also why do we need to be setting the WC session to "track events for unauthorized users"? If the only case we need it is when redirect to cart is enabled, then can't we just set `$deferred_conversion_events_transient_key` at a later point (WC will have set the session before `woocommerce_add_to_cart` is called)? There is a reason why WooCommerce doesn't always set the tracking cookies, and I think we should respect that, and not change this behaviour.

Reply:
> Also why do we need to be setting the WC session to "track events for unauthorized users"? If the only case we need it is when redirect to cart is enabled, then can't we just set $deferred_conversion_events_transient_key at a later point (WC will have set the session before woocommerce_add_to_cart is called)? There is a reason why WooCommerce doesn't always set the tracking cookies, and I think we should respect that, and not change this behaviour.

Reply:
> For an add to cart event we hook into `woocommerce_add_to_cart`, which will always be called before `wp_head`. How are we recording that if we don't have redirect to cart enabled? Does that mean we are still scheduling it for the next request? For example on a single product page when redirect is not enabled (or on a shop page if ajax and redirect are disabled). From my testing it is still being scheduled which makes it seem like the `did_action( 'wp_head' )` is not the right way to detect if we should enqueue or not.

The `woocommerce_add_to_cart` [enqueues the event](https://github.com/woocommerce/pinterest-for-woocommerce/blob/7d7f82c8bbf27c259bcf6f47e41d0443c80bf866/src/Tracking.php#L213), and since it is going to be called before `wp_head`, the enqueued event is always printed when `wp_head` is called.
If ajax is enabled and redirect is disabled we are [using JS to track an async ajax event](https://github.com/woocommerce/pinterest-for-woocommerce/blob/7d7f82c8bbf27c259bcf6f47e41d0443c80bf866/src/Tracking.php#L92).
If ajax and redirect are disabled, then the event is enqueued and printed in the same runtime (on `wp_head`).

> Also why do we need to be setting the WC session to "track events for unauthorized users"? If the only case we need it is when redirect to cart is enabled, then can't we just set `$deferred_conversion_events_transient_key` at a later point (WC will have set the session before `woocommerce_add_to_cart` is called)? There is a reason why WooCommerce doesn't always set the tracking cookies, and I think we should respect that, and not change this behaviour.
Correct. 
We included a change to set the `$deferred_conversion_events_transient_key` at a later point, since the only case is when redirect to cart is enabled.

Reply:
Thanks for the additional changes and clarifications. I still found one scenario where it doesn't look like it's handling it correctly. Can you clarify what the expected behaviour should be there?

##### Single product page add to cart (without redirect)
Regardless of the ajax setting, on the single product page it adds to the cart using a POST request to the current page. If I set a breakpoint in the function `hook_add_to_cart_event` and then add a product to the cart, I get the following values:
![image](https://user-images.githubusercontent.com/11388669/157826866-a25e55b6-1b41-4cbc-a792-0734b6e21a0d.png)

Since `wp_head` hasn't run yet it will queue the event and I'm seeing the transient being saved. So the event won't get tracked till I view another page.

Same scenario occurs when I disable ajax add to cart and try to add to cart from an archive/shop page. The event is also queued.

Reply:
> Single product page add to cart (without redirect)
Regardless of the ajax setting, on the single product page it adds to the cart using a POST request to the current page. If I set a breakpoint in the function hook_add_to_cart_event and then add a product to the cart, I get the following values:

Correct.
In this case the transient is saved in the function `hook_add_to_cart_event`. 

> So the event won't get tracked till I view another page.

When the product is added to the cart the page is reloaded and the event saved in the transient is printed and cleaned on `wp_head`. So, the event is tracked in the same runtime.

> Same scenario occurs when I disable ajax add to cart and try to add to cart from an archive/shop page. The event is also queued.

Same happens here. The event is captured in `hook_add_to_cart_event` and printed and cleaned on `wp_head` in the same runtime.